### PR TITLE
[hdf5] enable parallel support via OpenMPI

### PR DESCRIPTION
**Pull request by planetmarshall**
Specify library name and version:  **hdf5**

Adds support for Parallel HDF5 via Open MPI with the option `hdf5:parallel` (`False` by default)

Also added a test package for the C API as the existing test package will fail if `hdf5:enable_cxx` is `False`.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by planetmarshall** [comment]
>   /Users/jenkins/w/BuildSingleReference/.conan/data/hdf5/1.10.6/_/_/build/bcc183a7682ab4fd78cc3143f1eca64f5ef93041/source_subfolder/src/H5Fsuper.c:1397:16: error: implicit declaration of function 'H5O__fsinfo_set_version' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
>               if(H5O__fsinfo_set_version(f, &fsinfo) < 0)
>                  ^
>   1 error generated.

That's weird - haven't seen this error anywhere else and the function's definitely declared at the top of the file: https://github.com/HDFGroup/hdf5/blob/1.10/master/src/H5Fsuper.c#:~:text=herr_t%20H5O__fsinfo_set_version(H5F_t%20*f%2C%20H5O_fsinfo_t%20*fsinfo)%3B

Don't have a local Mac system but will see if I can reproduce on a github workflow.

**Comment by planetmarshall** [comment]
> > /Users/jenkins/w/BuildSingleReference/.conan/data/hdf5/1.10.6/_/_/build/bcc183a7682ab4fd78cc3143f1eca64f5ef93041/source_subfolder/src/H5Fsuper.c:1397:16: error: implicit declaration of function 'H5O__fsinfo_set_version' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
> > if(H5O__fsinfo_set_version(f, &fsinfo) < 0)
> > ^
> > 1 error generated.
> 
> That's weird - haven't seen this error anywhere else and the function's definitely declared at the top of the file: https://github.com/HDFGroup/hdf5/blob/1.10/master/src/H5Fsuper.c#:~:text=herr_t%20H5O__fsinfo_set_version(H5F_t%20*f%2C%20H5O_fsinfo_t%20*fsinfo)%3B
> 
> Couldn't reproduce it on my local system (Ubuntu 20.04, Clang-11) with `-Werror=implicit-function-declaration"`. Don't have a local Mac system but will see if I can reproduce on a github workflow.

Aha - it's specific to 1.10.6. Clang is right :)

File: recipes/hdf5/all/conanfile.py [hunk 1]
```diff
@@ -67,6 +68,11 @@ def configure(self):
              self.options.szip_encoding and \
              not self.options["szip"].enable_encoding:
             raise ConanInvalidConfiguration("encoding must be enabled in szip dependency (szip:enable_encoding=True)")
+        if self.options.parallel:
+            if self.options.enable_cxx:
+                raise ConanInvalidConfiguration("Parallel and C++ options are mutually exclusive")
+            if self.options.get_safe("threadsafe", False):
+                raise ConanInvalidConfiguration("Parallel and Threadsafe options are mutually exclusive")
```
Comment (uilianries) [inline recipes/hdf5/all/conanfile.py:75]:
```suggestion
        if self.options.parallel:
            if self.options.enable_cxx:
                raise ConanInvalidConfiguration("Parallel and C++ options are mutually exclusive")
            if self.options.get_safe("threadsafe", False):
                raise ConanInvalidConfiguration("Parallel and Threadsafe options are mutually exclusive")
```

Please, move it to `validate()` method

**Review by uilianries — APPROVED** [review]

File: recipes/hdf5/all/conanfile.py [hunk 2]
```diff
@@ -72,6 +73,11 @@ def validate(self):
         if hasattr(self, "settings_build") and tools.cross_building(self, skip_x64_x86=True):
             # While building it runs some executables like H5detect
             raise ConanInvalidConfiguration("Current recipe doesn't support cross-building (yet)")
+        if self.options.parallel:
+            if self.options.enable_cxx:
+                raise ConanInvalidConfiguration("Parallel and C++ options are mutually exclusive")
```
Comment (SSE4) [inline recipes/hdf5/all/conanfile.py:78]: this is unfortunate (?)

File: recipes/hdf5/all/conanfile.py [hunk 3]
```diff
@@ -72,6 +73,11 @@ def validate(self):
...
+            if self.options.get_safe("threadsafe", False):
+                raise ConanInvalidConfiguration("Parallel and Threadsafe options are mutually exclusive")
```
Comment (SSE4) [inline recipes/hdf5/all/conanfile.py:80]: as this... parallel but not thread safe, wut?

**Review by SSE4 — APPROVED** [review]

File: recipes/hdf5/all/conanfile.py (hunk 3 above)
Comment (planetmarshall) [reply recipes/hdf5/all/conanfile.py:80]: Yes it's a bit confusing. They refer to two different things in HDF5
* `threadsafe` is for parallel access in shared memory systems - it basically just serializes access using a mutex.
* `parallel` refers to the HDF5 Parallel library which allows true parallel access using process level parallelism via MPI, so no shared memory.

File: recipes/hdf5/all/conanfile.py (hunk 2 above)
Comment (planetmarshall) [reply recipes/hdf5/all/conanfile.py:78]: Yes the H5CPP library and HDF5 Parallel are an "untested configuration", so HDF5 makes them mutually exclusive by default - though it is possible to use them together in an "at your own risk" configuration.

It is possible however to use HDF5 as a low-level implementation to support higher level C++ implementations (ie, https://github.com/conan-io/conan-center-index/pull/6239). Hence my interest in this PR :)

**Review by prince-chrismc — APPROVED** [review]

**Pull request merged**